lib/kube/proxy/resource_filters.go

```diff
@@ -0,0 +1,674 @@
+		return err == nil, false, nil
+
+	case *networkingv1.IngressList:
+		o.Items = pointerArrayToArray(
+			filterResourceList(
+				d.kind,
+				arrayToPointerArray(o.Items), d.allowedResources, d.deniedResources, d.log),
+		)
+		return len(o.Items) > 0, true, nil
+
+	case *metav1.Table:
+		_, err := d.filterMetaV1Table(o, d.allowedResources, d.deniedResources)
+		if err != nil {
+			return false, false, trace.Wrap(err)
+		}
+		return len(o.Rows) > 0, true, nil
+
+	default:
+		// It's important default types are never blindly forwarded or protocol
+		// extensions could result in information disclosures.
+		return false, false, trace.BadParameter("unexpected type received; got %T", obj)
+	}
+}
+
+// decode decodes the buffer into the appropriate type if the responseCode
+// belongs to the range 200(OK)-206(PartialContent).
+// If it does not belong, it returns the buffer unchanged since it contains
+// an error message from the Kubernetes API server and it's safe to return
+// it back to the user.
+func (d *resourceFilterer) decode(buffer []byte) (runtime.Object, []byte, error) {
+	switch {
+	case d.responseCode == http.StatusSwitchingProtocols:
+		// no-op, we've been upgraded
+		return nil, buffer, nil
+	case d.responseCode < http.StatusOK /* 200 */ || d.responseCode > http.StatusPartialContent /* 206 */ :
+		// calculate an unstructured error from the response which the Result object may use if the caller
+		// did not return a structured error.
+		// Logic from: https://github.com/kubernetes/client-go/blob/58ff029093df37cad9fa28778a37f11fa495d9cf/rest/request.go#L1040
+		return nil, buffer, nil
+	default:
+		out, err := decodeAndSetGVK(d.decoder, buffer)
+		return out, nil, trace.Wrap(err)
+	}
+}
+
+// decodePartialObjectMetadata decodes the metav1.PartialObjectMetadata present
+// in the metav1.TableRow entry. This information comes from server side and
+// includes the resource name and namespace as a structured object.
+func (d *resourceFilterer) decodePartialObjectMetadata(row *metav1.TableRow) error {
+	if row.Object.Object != nil {
+		return nil
+	}
+	var err error
+	// decode only if row.Object.Object was not decoded before.
+	row.Object.Object, err = decodeAndSetGVK(d.decoder, row.Object.Raw)
+	return trace.Wrap(err)
+}
+
+// encode encodes the filtered object into the io.Writer using the same
+// content-type.
+func (d *resourceFilterer) encode(obj runtime.Object, w io.Writer) error {
+	return trace.Wrap(d.encoder.Encode(obj, w))
+}
+
+// filterResourceList excludes resources the user should not have access to.
+func filterResourceList[T kubeObjectInterface](kind string, originalList []T, allowed, denied []types.KubernetesResource, log logrus.FieldLogger) []T {
+	filteredList := make([]T, 0, len(originalList))
+	for _, resource := range originalList {
+		if err := filterResource(kind, resource, allowed, denied); err == nil {
+			filteredList = append(filteredList, resource)
+		} else if !trace.IsAccessDenied(err) {
```

Comment:
if performance is a concern at all I'd recommend avoiding trace errors as much as possible in filter loops where it's not actually an error, since the error creation will collect a stack trace for every resource that gets filtered out